File: jmlparser-jml-tools/src/main/java/com/github/jmlparser/jml2java/Jml2JavaTranslator.java
package com.github.jmlparser.jml2java;

import com.beust.ah.A;
import com.github.javaparser.ast.Modifier;
import com.github.javaparser.ast.body.VariableDeclarator;
import com.github.javaparser.ast.expr.*;
import com.github.javaparser.ast.jml.expr.JmlLetExpr;
import com.github.javaparser.ast.jml.expr.JmlMultiCompareExpr;
import com.github.javaparser.ast.jml.expr.JmlQuantifiedExpr;
import com.github.javaparser.ast.stmt.*;
import com.github.javaparser.ast.type.PrimitiveType;
import com.github.javaparser.ast.type.VarType;
import com.github.javaparser.ast.visitor.GenericVisitorAdapter;
import com.github.jmlparser.utils.JMLUtils;
import com.github.jmlparser.utils.Pattern;

import java.util.concurrent.atomic.AtomicInteger;

/**
 * @author Alexander Weigl
 * @version 1 (04.10.22)
 */
public class Jml2JavaTranslator {
    private final AtomicInteger counter = new AtomicInteger();

    public Expression accept(Expression e, BlockStmt arg) {
        if (Jml2JavaFacade.containsJmlExpression(e)) {
            return e.accept(new Jml2JavaVisitor(), arg);
        }
        return e; // createAssignmentAndAdd(e, arg);
    }

    private Expression createAssignmentAndAdd(Expression e, BlockStmt arg) {
        arg.addAndGetStatement(createAssignmentFor(e));
        return new NameExpr(getTargetForAssignment());
    }

    private Statement createAssignmentFor(Expression e) {
        var decl = new VariableDeclarationExpr(
                new VariableDeclarator(new VarType(),
                        newTargetForAssignment(), e));
        decl.addModifier(Modifier.DefaultKeyword.FINAL);
        return new ExpressionStmt(decl);
    }

    private SimpleName getTargetForAssignment() {
        return new SimpleName("_gen_" + counter.get());
    }

    private SimpleName newTargetForAssignment() {
        counter.getAndIncrement();
        return getTargetForAssignment();
    }

    public Expression findPredicate(JmlQuantifiedExpr n) {
        return n.getExpressions().get(n.getExpressions().size() - 1);
    }

    public static Expression findBound(JmlQuantifiedExpr n) {
        if (n.getExpressions().size() == 2)
            return n.getExpressions().get(0);
        else if (n.getExpressions().size() == 1)
            if (n.getExpressions().get(0) instanceof BinaryExpr be)
                return be.getLeft();
        throw new IllegalArgumentException("Could not determine bound.");

    }

    public static Expression findLowerBound(JmlQuantifiedExpr n, String variable) {
        if (n.getExpressions().size() == 3) return n.getExpressions().get(0);

        var e = findBound(n);

        if (e instanceof JmlMultiCompareExpr mc) {
            if (mc.getExpressions().size() == 3 &&
                    mc.getExpressions().get(1) instanceof NameExpr v &&
                    v.getNameAsString().equals(variable)) {
                if (mc.getOperators().get(0) == BinaryExpr.Operator.LESS_EQUALS)
                    return mc.getExpressions().get(0);
                if (mc.getOperators().get(0) == BinaryExpr.Operator.LESS)
                    return new BinaryExpr(mc.getExpressions().get(0), new IntegerLiteralExpr(1), BinaryExpr.Operator.PLUS);
                throw new IllegalStateException();
            }
        }

        Expression ph = new NameExpr("_");
        {
            var be = new BinaryExpr(new NameExpr(variable), ph, BinaryExpr.Operator.GREATER_EQUALS);
            var pattern = new Pattern<>(be);
            pattern.addPlaceholder(ph, "min");
            var result = pattern.find(n);
            if (result != null)
                return (Expression) result.get("min");
        }

        {
            var be = new BinaryExpr(new NameExpr(variable), ph, BinaryExpr.Operator.GREATER);
            var pattern = new Pattern<>(be);
            pattern.addPlaceholder(ph, "min");
            var result = pattern.find(n);
            if (result != null)
                return new BinaryExpr((Expression) result.get("min"), new IntegerLiteralExpr(1), BinaryExpr.Operator.PLUS);
        }

        {
            var be = new BinaryExpr(ph, new NameExpr(variable), BinaryExpr.Operator.LESS_EQUALS);
            var pattern = new Pattern<>(be);
            pattern.addPlaceholder(ph, "min");
            var result = pattern.find(n);
            if (result != null)
                return (Expression) result.get("min");
        }

        {
            var be = new BinaryExpr(ph, new NameExpr(variable), BinaryExpr.Operator.LESS);
            var pattern = new Pattern<>(be);
            pattern.addPlaceholder(ph, "min");
            var result = pattern.find(n);
            if (result != null)
                return new BinaryExpr((Expression) result.get("min"), new IntegerLiteralExpr(1), BinaryExpr.Operator.PLUS);
        }

        return null;
    }

    public Expression findUpperBound(JmlQuantifiedExpr n, String variable) {
        if (n.getExpressions().size() == 3) return n.getExpressions().get(1);

        var e = findBound(n);

        if (e instanceof JmlMultiCompareExpr mc) {
            if (mc.getExpressions().size() == 3 &&
                    mc.getExpressions().get(1) instanceof NameExpr v &&
                    v.getNameAsString().equals(variable)) {
                if (mc.getOperators().get(0) == BinaryExpr.Operator.LESS_EQUALS)
                    return mc.getExpressions().get(0);
                if (mc.getOperators().get(0) == BinaryExpr.Operator.LESS)
                    return new BinaryExpr(mc.getExpressions().get(0), new IntegerLiteralExpr(1), BinaryExpr.Operator.PLUS);
                throw new IllegalStateException();
            }
        }

        Expression ph = new NameExpr("_");
        {
            var be = new BinaryExpr(new NameExpr(variable), ph, BinaryExpr.Operator.GREATER_EQUALS);
            var pattern = new Pattern<>(be);
            pattern.addPlaceholder(ph, "min");
            var result = pattern.find(n);
            if (result != null)
                return (Expression) result.get("min");
        }

        {
            var be = new BinaryExpr(new NameExpr(variable), ph, BinaryExpr.Operator.GREATER);
            var pattern = new Pattern<>(be);
            pattern.addPlaceholder(ph, "min");
            var result = pattern.find(n);
            if (result != null)
                return new BinaryExpr((Expression) result.get("min"), new IntegerLiteralExpr(1), BinaryExpr.Operator.PLUS);
        }

        {
            var be = new BinaryExpr(ph, new NameExpr(variable), BinaryExpr.Operator.LESS_EQUALS);
            var pattern = new Pattern<>(be);
            pattern.addPlaceholder(ph, "min");
            var result = pattern.find(n);
            if (result != null)
                return (Expression) result.get("min");
        }

        {
            var be = new BinaryExpr(ph, new NameExpr(variable), BinaryExpr.Operator.LESS);
            var pattern = new Pattern<>(be);
            pattern.addPlaceholder(ph, "min");
            var result = pattern.find(n);
            if (result != null)
                return new BinaryExpr((Expression) result.get("min"), new IntegerLiteralExpr(1), BinaryExpr.Operator.PLUS);
        }

        return null;
    }

    private final class Jml2JavaVisitor extends GenericVisitorAdapter<Expression, BlockStmt> {
        @Override
        public Expression visit(JmlQuantifiedExpr n, BlockStmt arg) {
            if (n.getBinder() == JmlQuantifiedExpr.JmlDefaultBinder.FORALL)
                return visitForall(n, arg);
            if (n.getBinder() == JmlQuantifiedExpr.JmlDefaultBinder.EXISTS)
                return visitExists(n, arg);

            throw new IllegalArgumentException("Unsupport quantifier " + n.getBinder());
        }

        private Expression visitForall(JmlQuantifiedExpr n, BlockStmt arg) {
            assert n.getVariables().size() == 1;
            var rvalue = newSymbol("result_");
            var highVar = newSymbol("high_");
            var predVar = newSymbol("pred_");

            arg.addAndGetStatement(
                    new ExpressionStmt(new VariableDeclarationExpr(
                            new VariableDeclarator(
                                    new PrimitiveType(PrimitiveType.Primitive.BOOLEAN),
                                    rvalue, new BooleanLiteralExpr(true)))));

            var variable = n.getVariables().get(0);
            var lowCode = Jml2JavaFacade.translate(findLowerBound(n, variable.getNameAsString()));
            arg.addAndGetStatement(
                    new ExpressionStmt(new VariableDeclarationExpr(
                            new VariableDeclarator(
                                    variable.getType().clone(), variable.getNameAsString()))));
            lowCode.a.addAndGetStatement(
                    new ExpressionStmt(
                            new AssignExpr(
                                    new NameExpr(variable.getNameAsString()), lowCode.b,
                                    AssignExpr.Operator.ASSIGN)));
            arg.addAndGetStatement(lowCode.a);

            var highCode = Jml2JavaFacade.translate(findUpperBound(n, variable.getNameAsString()));
            arg.addAndGetStatement(
                    new ExpressionStmt(new VariableDeclarationExpr(
                            new VariableDeclarator(
                                    variable.getType().clone(), highVar))));
            highCode.a.addAndGetStatement(
                    new ExpressionStmt(
                            new AssignExpr(
                                    new NameExpr(highVar), highCode.b,
                                    AssignExpr.Operator.ASSIGN)));
            arg.addAndGetStatement(highCode.a.clone());

            var predCode = Jml2JavaFacade.translate(findPredicate(n));

            WhileStmt whileStmt = arg.addAndGetStatement(new WhileStmt());
            var lessThanExpr = new BinaryExpr(new NameExpr(variable.getNameAsString()),
                    new NameExpr(highVar), BinaryExpr.Operator.LESS);
            whileStmt.setCondition(
                    new BinaryExpr(lessThanExpr, new NameExpr(rvalue), BinaryExpr.Operator.AND));
            var body = new BlockStmt();

            body.addAndGetStatement(
                    new ExpressionStmt(new VariableDeclarationExpr(
                            new VariableDeclarator(new PrimitiveType(PrimitiveType.Primitive.BOOLEAN), predVar))));
            predCode.a.addAndGetStatement(
                    new ExpressionStmt(
                            new AssignExpr(
                                    new NameExpr(predVar), predCode.b,
                                    AssignExpr.Operator.ASSIGN)));
            body.addAndGetStatement(predCode.a);

            final var assignPred = new ExpressionStmt(new AssignExpr(new NameExpr(rvalue),
                    new BooleanLiteralExpr(true), AssignExpr.Operator.ASSIGN));
            body.addAndGetStatement(
                    new IfStmt(new NameExpr(predVar), assignPred, null));
            body.addAndGetStatement(highCode.a.clone());

            whileStmt.setBody(body);
            return new NameExpr(rvalue);
        }

        private String newSymbol(String prefix) {
            return prefix + counter.getAndIncrement();
        }

        private Expression visitExists(JmlQuantifiedExpr n, BlockStmt arg) {
            return new UnaryExpr(visitForall(n, arg), UnaryExpr.Operator.LOGICAL_COMPLEMENT);
        }


        @Override
        public Expression visit(JmlLetExpr n, BlockStmt arg) {
            var inner = new BlockStmt();
            SimpleName target = newTargetForAssignment();
            var type = n.getBody().calculateResolvedType();
            arg.addAndGetStatement(
                    new ExpressionStmt(new VariableDeclarationExpr(JMLUtils.resolvedType2Type(type),
                            target.asString())));
            inner.addAndGetStatement(new ExpressionStmt(n.getVariables()));
            var e = accept(n.getBody(), inner);
            arg.addAndGetStatement(inner);
            inner.addAndGetStatement(new AssignExpr(new NameExpr(target.asString()), e, AssignExpr.Operator.ASSIGN));
            return new NameExpr(target.asString());
        }

        @Override
        public Expression visit(BinaryExpr n, BlockStmt arg) {
            var left = accept(n.getLeft(), arg);
            var right = accept(n.getRight(), arg);
            switch (n.getOperator()) {
                case IMPLICATION:
                    return new BinaryExpr(
                            new UnaryExpr(new EnclosedExpr(left), UnaryExpr.Operator.LOGICAL_COMPLEMENT),
                            right,
                            BinaryExpr.Operator.OR);
                case RIMPLICATION:
                    return
                            new BinaryExpr(
                                    left,
                                    new UnaryExpr(new EnclosedExpr(right), UnaryExpr.Operator.LOGICAL_COMPLEMENT),
                                    BinaryExpr.Operator.OR);
                case EQUIVALENCE:
                    return new BinaryExpr(left, right, BinaryExpr.Operator.EQUALS);
                case SUBTYPE:
                case SUB_LOCK:
                case SUB_LOCKE:
                    throw new IllegalArgumentException("Unsupported operators.");
                default:
                    return createAssignmentAndAdd(
                            new BinaryExpr(left, right, n.getOperator()),
                            arg);
            }
        }

        @Override
        public Expression visit(ArrayAccessExpr n, BlockStmt arg) {
            return super.visit(n, arg);
        }

        @Override
        public Expression visit(ArrayCreationExpr n, BlockStmt arg) {
            return super.visit(n, arg);
        }

        @Override
        public Expression visit(ArrayInitializerExpr n, BlockStmt arg) {
            return super.visit(n, arg);
        }

        @Override
        public Expression visit(AssignExpr n, BlockStmt arg) {
            return super.visit(n, arg);
        }


    }
}

File: tools/jml2java/src/main/kotlin/io/github/jmltoolkit/jml2java/Jml2JavaTranslator.kt
package io.github.jmltoolkit.jml2java

import com.github.javaparser.ast.Modifier
import com.github.javaparser.ast.body.Parameter
import com.github.javaparser.ast.body.VariableDeclarator
import com.github.javaparser.ast.expr.*
import com.github.javaparser.ast.jml.expr.JmlLetExpr
import com.github.javaparser.ast.jml.expr.JmlMultiCompareExpr
import com.github.javaparser.ast.jml.expr.JmlQuantifiedExpr
import com.github.javaparser.ast.stmt.*
import com.github.javaparser.ast.type.PrimitiveType
import com.github.javaparser.ast.type.VarType
import com.github.javaparser.ast.visitor.GenericVisitorAdapter
import com.github.javaparser.resolution.types.ResolvedType
import io.github.jmltoolkit.utils.JMLUtils
import io.github.jmltoolkit.utils.Pattern
import java.util.concurrent.atomic.AtomicInteger

/**
 * @author Alexander Weigl
 * @version 1 (04.10.22)
 */
class Jml2JavaTranslator {
    private val counter = AtomicInteger()

    fun accept(e: Expression?, arg: BlockStmt): Expression? {
        if (Jml2JavaFacade.containsJmlExpression(e)) {
            return e!!.accept(Jml2JavaVisitor(), arg)
        }
        return e // createAssignmentAndAdd(e, arg);
    }

    private fun createAssignmentAndAdd(e: Expression, arg: BlockStmt): Expression {
        arg.addAndGetStatement(createAssignmentFor(e))
        return NameExpr(targetForAssignment)
    }

    private fun createAssignmentFor(e: Expression): Statement {
        val decl = VariableDeclarationExpr(
            VariableDeclarator(
                VarType(),
                newTargetForAssignment(), e
            )
        )
        decl.addModifier(Modifier.DefaultKeyword.FINAL)
        return ExpressionStmt(decl)
    }

    private val targetForAssignment: SimpleName
        get() = SimpleName("_gen_" + counter.get())

    private fun newTargetForAssignment(): SimpleName {
        counter.getAndIncrement()
        return targetForAssignment
    }

    fun findPredicate(n: JmlQuantifiedExpr): Expression {
        return n.expressions[n.expressions.size - 1]
    }

    fun findUpperBound(n: JmlQuantifiedExpr, variable: String): Expression? {
        if (n.expressions.size == 3) return n.expressions[1]

        val e = findBound(n)

        if (e is JmlMultiCompareExpr) {
            if (e.expressions.size == 3 &&
                (e.expressions[1] as? NameExpr)?.nameAsString == variable
            ) {
                if (e.operators[0] == BinaryExpr.Operator.LESS_EQUALS) return e.expressions[0]
                if (e.operators[0] == BinaryExpr.Operator.LESS) return BinaryExpr(
                    e.expressions[0],
                    IntegerLiteralExpr("1"),
                    BinaryExpr.Operator.PLUS
                )
                throw IllegalStateException()
            }
        }

        val ph: Expression = NameExpr("_")
        run {
            val be = BinaryExpr(NameExpr(variable), ph, BinaryExpr.Operator.GREATER_EQUALS)
            val pattern = Pattern(be)
            pattern.addPlaceholder(ph, "min")
            val result = pattern.find(n)
            if (result != null) return result["min"] as Expression?
        }

        run {
            val be = BinaryExpr(NameExpr(variable), ph, BinaryExpr.Operator.GREATER)
            val pattern = Pattern(be)
            pattern.addPlaceholder(ph, "min")
            val result = pattern.find(n)
            if (result != null) return BinaryExpr(
                result["min"] as Expression?,
                IntegerLiteralExpr("1"),
                BinaryExpr.Operator.PLUS
            )
        }

        run {
            val be = BinaryExpr(ph, NameExpr(variable), BinaryExpr.Operator.LESS_EQUALS)
            val pattern = Pattern(be)
            pattern.addPlaceholder(ph, "min")
            val result = pattern.find(n)
            if (result != null) return result["min"] as Expression?
        }

        run {
            val be = BinaryExpr(ph, NameExpr(variable), BinaryExpr.Operator.LESS)
            val pattern = Pattern(be)
            pattern.addPlaceholder(ph, "min")
            val result = pattern.find(n)
            if (result != null) return BinaryExpr(
                result["min"] as Expression?,
                IntegerLiteralExpr("1"),
                BinaryExpr.Operator.PLUS
            )
        }

        return null
    }

    private inner class Jml2JavaVisitor : GenericVisitorAdapter<Expression, BlockStmt>() {
        override fun visit(n: JmlQuantifiedExpr, arg: BlockStmt): Expression {
            if (n.binder === JmlQuantifiedExpr.JmlDefaultBinder.FORALL) return visitForall(n, arg)
            if (n.binder === JmlQuantifiedExpr.JmlDefaultBinder.EXISTS) return visitExists(n, arg)

            throw IllegalArgumentException("Unsupport quantifier " + n.binder)
        }

        private fun visitForall(n: JmlQuantifiedExpr, arg: BlockStmt): Expression {
            assert(n.variables.size == 1)
            val rvalue = newSymbol("result_")
            val highVar = newSymbol("high_")
            val predVar = newSymbol("pred_")

            arg.addAndGetStatement(
                ExpressionStmt(
                    VariableDeclarationExpr(
                        VariableDeclarator(
                            PrimitiveType(PrimitiveType.Primitive.BOOLEAN),
                            rvalue, BooleanLiteralExpr(true)
                        )
                    )
                )
            )

            val variable: Parameter = n.variables[0]
            val lowCode = Jml2JavaFacade.translate(findLowerBound(n, variable.nameAsString))
            arg.addAndGetStatement(
                ExpressionStmt(
                    VariableDeclarationExpr(
                        VariableDeclarator(
                            variable.type.clone(), variable.nameAsString
                        )
                    )
                )
            )
            lowCode.a!!.addAndGetStatement(
                ExpressionStmt(
                    AssignExpr(
                        NameExpr(variable.nameAsString), lowCode.b,
                        AssignExpr.Operator.ASSIGN
                    )
                )
            )
            arg.addAndGetStatement(lowCode.a)

            val highCode = Jml2JavaFacade.translate(findUpperBound(n, variable.nameAsString))
            arg.addAndGetStatement(
                ExpressionStmt(
                    VariableDeclarationExpr(
                        VariableDeclarator(
                            variable.type.clone(), highVar
                        )
                    )
                )
            )
            highCode.a!!.addAndGetStatement(
                ExpressionStmt(
                    AssignExpr(
                        NameExpr(highVar), highCode.b,
                        AssignExpr.Operator.ASSIGN
                    )
                )
            )
            arg.addAndGetStatement(highCode.a!!.clone())

            val predCode = Jml2JavaFacade.translate(findPredicate(n))

            val whileStmt = arg.addAndGetStatement(WhileStmt())
            val lessThanExpr = BinaryExpr(
                NameExpr(variable.nameAsString),
                NameExpr(highVar), BinaryExpr.Operator.LESS
            )
            whileStmt.setCondition(
                BinaryExpr(lessThanExpr, NameExpr(rvalue), BinaryExpr.Operator.AND)
            )
            val body = BlockStmt()

            body.addAndGetStatement(
                ExpressionStmt(
                    VariableDeclarationExpr(
                        VariableDeclarator(PrimitiveType(PrimitiveType.Primitive.BOOLEAN), predVar)
                    )
                )
            )
            predCode.a!!.addAndGetStatement(
                ExpressionStmt(
                    AssignExpr(
                        NameExpr(predVar), predCode.b,
                        AssignExpr.Operator.ASSIGN
                    )
                )
            )
            body.addAndGetStatement(predCode.a)

            val assignPred = ExpressionStmt(
                AssignExpr(
                    NameExpr(rvalue),
                    BooleanLiteralExpr(true), AssignExpr.Operator.ASSIGN
                )
            )
            body.addAndGetStatement(
                IfStmt(NameExpr(predVar), assignPred, null)
            )
            body.addAndGetStatement(highCode.a!!.clone())

            whileStmt.setBody(body)
            return NameExpr(rvalue)
        }

        private fun newSymbol(prefix: String): String {
            return prefix + counter.getAndIncrement()
        }

        private fun visitExists(n: JmlQuantifiedExpr, arg: BlockStmt): Expression {
            return UnaryExpr(visitForall(n, arg), UnaryExpr.Operator.LOGICAL_COMPLEMENT)
        }


        override fun visit(n: JmlLetExpr, arg: BlockStmt): Expression {
            val inner = BlockStmt()
            val target = newTargetForAssignment()
            val type: ResolvedType = n.body.calculateResolvedType()
            arg.addAndGetStatement<ExpressionStmt>(
                ExpressionStmt(
                    VariableDeclarationExpr(
                        JMLUtils.resolvedType2Type(type),
                        target.asString()
                    )
                )
            )
            inner.addAndGetStatement(ExpressionStmt(n.variables))
            val e = accept(n.body, inner)
            arg.addAndGetStatement(inner)
            inner.addAndGetStatement(AssignExpr(NameExpr(target.asString()), e, AssignExpr.Operator.ASSIGN))
            return NameExpr(target.asString())
        }

        override fun visit(n: BinaryExpr, arg: BlockStmt): Expression {
            val left = accept(n.left, arg)
            val right = accept(n.right, arg)
            return when (n.operator) {
                BinaryExpr.Operator.IMPLICATION -> BinaryExpr(
                    UnaryExpr(
                        EnclosedExpr(left),
                        UnaryExpr.Operator.LOGICAL_COMPLEMENT
                    ),
                    right,
                    BinaryExpr.Operator.OR
                )

                BinaryExpr.Operator.RIMPLICATION -> BinaryExpr(
                    left,
                    UnaryExpr(
                        EnclosedExpr(right),
                        UnaryExpr.Operator.LOGICAL_COMPLEMENT
                    ),
                    BinaryExpr.Operator.OR
                )

                BinaryExpr.Operator.EQUIVALENCE -> BinaryExpr(
                    left,
                    right,
                    BinaryExpr.Operator.EQUALS
                )

                BinaryExpr.Operator.SUBTYPE, BinaryExpr.Operator.SUB_LOCK, BinaryExpr.Operator.SUB_LOCKE -> throw IllegalArgumentException(
                    "Unsupported operators."
                )

                else -> createAssignmentAndAdd(
                    BinaryExpr(left, right, n.operator),
                    arg
                )
            }
        }

        override fun visit(n: ArrayAccessExpr, arg: BlockStmt): Expression {
            return super.visit(n, arg)
        }

        override fun visit(n: ArrayCreationExpr, arg: BlockStmt): Expression {
            return super.visit(n, arg)
        }

        override fun visit(n: ArrayInitializerExpr, arg: BlockStmt): Expression {
            return super.visit(n, arg)
        }

        override fun visit(n: AssignExpr, arg: BlockStmt): Expression {
            return super.visit(n, arg)
        }
    }

    companion object {
        fun findBound(n: JmlQuantifiedExpr): Expression {
            if (n.expressions.size == 2) return n.expressions[0]
            else
                if (n.expressions.size == 1 && n.expressions.first() is BinaryExpr)
                    return n.expressions.first()
            throw IllegalArgumentException("Could not determine bound.")
        }

        fun findLowerBound(n: JmlQuantifiedExpr, variable: String): Expression? {
            if (n.expressions.size == 3) return n.expressions[0]

            val e = findBound(n)

            if (e is JmlMultiCompareExpr) {
                if (e.expressions.size == 3 &&
                    (e.expressions[1] as? NameExpr)?.nameAsString == variable
                ) {
                    if (e.operators[0] == BinaryExpr.Operator.LESS_EQUALS) return e.expressions[0]
                    if (e.operators[0] == BinaryExpr.Operator.LESS) return BinaryExpr(
                        e.expressions[0], IntegerLiteralExpr("1"), BinaryExpr.Operator.PLUS
                    )
                    throw IllegalStateException()
                }
            }

            val ph: Expression = NameExpr("_")
            run {
                val be = BinaryExpr(NameExpr(variable), ph, BinaryExpr.Operator.GREATER_EQUALS)
                val pattern = Pattern(be)
                pattern.addPlaceholder(ph, "min")
                val result = pattern.find(n)
                if (result != null) return result["min"] as Expression?
            }

            run {
                val be = BinaryExpr(NameExpr(variable), ph, BinaryExpr.Operator.GREATER)
                val pattern = Pattern(be)
                pattern.addPlaceholder(ph, "min")
                val result = pattern.find(n)
                if (result != null) return BinaryExpr(
                    result["min"] as Expression?,
                    IntegerLiteralExpr("1"),
                    BinaryExpr.Operator.PLUS
                )
            }

            run {
                val be = BinaryExpr(ph, NameExpr(variable), BinaryExpr.Operator.LESS_EQUALS)
                val pattern = Pattern(be)
                pattern.addPlaceholder(ph, "min")
                val result = pattern.find(n)
                if (result != null) return result["min"] as Expression?
            }

            run {
                val be = BinaryExpr(ph, NameExpr(variable), BinaryExpr.Operator.LESS)
                val pattern = Pattern(be)
                pattern.addPlaceholder(ph, "min")
                val result = pattern.find(n)
                if (result != null) return BinaryExpr(
                    result["min"] as Expression?,
                    IntegerLiteralExpr("1"),
                    BinaryExpr.Operator.PLUS
                )
            }

            return null
        }
    }
}
